README
uGopherServer, the universal gopher server
See LICENSE for copyright information

uGopherServer is a modern, featureful and relatively secure gopher
daemon built to be run under inetd, available under the 2-clause
BSD license. Binaries are available under the combination of the
2-clause BSD license, the SSLeay license and the OpenSSL license.

uGS accepts the environment variables "REMOTE_ADDR" and "REMOTE_HOST"
with the same semantics so that if it is being run behind some kind
of proxy, it is possible for that proxy, with a counterpart
shielding uGS, to pass the address through to uGS.

It is based on the proven codebase of Gophernicus 2014, and includes
numerous additions for TLS support, including user privilege droppage,
and obviously, TLS support. You MUST have OpenSSL available in
/usr/lib and /usr/include for it to compile.

Command line options:
    -h hostname   Change server hostname (FQDN)      [$HOSTNAME]
    -p port       Change server port                 [70]
                  Note that for a -S instance, you must specify a -p
                  above 100000 (it is suggested, if listening on
                  105, formerly the CSO/ph port, to use -p 100105) 
    -r root       Change gopher root                 [/var/gopher]
    -t type       Change default gopher filetype     [0]
    -g mapfile    Change gophermap file              [gophermap]
    -a tagfile    Change gophertag file              [gophertag]
    -c cgidir     Change CGI script directory        [/cgi-bin/]
    -u userdir    Change users personal gopherspace  [public_gopher]
    -l logfile    Log to Apache-compatible combined format logfile

    -H headerext  Change extension of header file    [Ghdr]
    -T tagext     Change extension of tag file       [Gtag]
    -F footerext  Change extension of footer file    [Gftr]

    -w width      Change default page width          [70]
                  Many sites would benefit from boosting this to 132.

    -o charset    Change default output charset      [US-ASCII]

    -s seconds    Session timeout in seconds         [1800]
    -i hits       Maximum hits until throttling      [4096]
    -k kbytes     Maximum transfer until throttling  [4194304]

    -f filterdir  Specify directory for output filters
    -e ext=type   Map file extension to gopher filetype
    -R old=new    Rewrite the beginning of a selector

    -D text|file  Set or load server description for caps.txt
    -L text|file  Set or load server location for caps.txt
    -A admin      Set admin email for caps.txt

    -S certkeypem Enable SSL/TLS protection          [disabled]
                  Requires a gophernicus compiled with OpenSSL
                  support. The argument must be specified and
                  is a path, preferably absolute, to your SSL
                  certificate. Must be readable by the user
                  the process runs as before going privdrop.
    -K cipherlist Specify custom cipherlist. No effect unless
                  -S. Normally you don't need to.

    -U droptouser Drop privileges to this username   [<none>]
                  after reading certificates.
    -G group      Drop privileges to this group      [group of -U]
                  after reading certificates.
    The above two options are unnecessary if running plaintext,
    however if the daemon is started as root, you must use them
    or disable root checking (not recommended.)

    -nv           Disable virtual hosting
    -nl           Disable parent directory links
    -nh           Disable menu header (title)
    -nf           Disable menu footer
    -nd           Disable dates and filesizes in menus
    -nc           Disable file content detection
    -no           Disable charset conversion for output
    -nq           Disable HTTP-style query strings (?query)
    -ns           Disable logging to syslog
    -na           Disable autogenerated caps.txt
    -nm           Disable shared memory use (for debugging)
    -nr           Disable root user checking (for debugging)

    -d            Debug to syslog (not for production use)
    -b            Display the BSD license
    -?            Display this help
 

Setting up a gopher site
========================

After succesfully installing uGopherServer (see INSTALL) you need to set
up the gopher root directory. By default uGopherServer serves documents
from /var/gopher so start by creating that directory and making sure
it's world-readable. Then, simply add files and directories under your
root, fire up a gopher browser (Firefox, Lynx) and open up this URL:

gopher://<HOSTNAME>/   (where <HOSTNAME> is your servers hostname)

That's it, your first gopher site is now up and running. If the links
on the root menu don't work, make sure you are using the -h <HOSTNAME>
parameter in your inetd.conf (with a valid resolveable hostname
instead of <HOSTNAME> - see INSTALL).


Security
========

Gophernicus, on which uGopherServer is based, was written with high
security in mind. There should be no buffer overflows or memory
allocation problems; it should be safe to run a publicly available
gopher server with uGopherServer.

However, the security settings (which are non-changeable) are so strict
that you need to keep one thing in mind. uGopherServer will only serve
world-readable content. Being readable by the server process is not
enough, all files and directories MUST be world-readable or they are
simply hidden from all listings and denied if a client asks for them.


Gophermaps
==========

By default all gopher menus are automatically generated from the
content of the directory being viewed. If you want to have
informational text along with the files, or if you want to completely
replace the generated menu with your own you need to take a look at
gophermaps. See the file README.gophermap for more information.

Menus can be writ piecemeal in part if you desire to have a preamble
before every menu item in uGopherServer. foo.Ghdr will be a
preamble to the item, and foo.Gftr will be the postamble.

Gophertags
==========

A gophertag file can be used to virtually rename a directory. Let's
assume that you have a directory called "foo" somewhere - it will
be listed as "foo" in all automatically generated menus. Now if you
create a file foo/gophertag and put the text "bar" into it the menus
will show "bar" but the links will still point to "foo". This is
useful for creating descriptive names for directories without
littering the file system with spaces and weird characters.

This feature is also supported for files now - create a foo.Gtag
to rename foo to the contents of the first line of foo.Gtag.

Personal gopherspaces
=====================

Gophernicus, and by extension, uGopherServer supports user
personal gopherspaces. If a user has a world-readable directory
called public_gopher/ in their home, a request for

   gopher://<HOSTNAME>/1/~user/

will serve documents from that directory.


Virtual hosting
===============

uGS supports virtual hosting, or serving more than one logical
domain using the same IP address. Since gopher (RFC1436) doesn't
support virtual hosting this requires some clever (but mostly
invisible) hacks.

To enable virtual hosting create one or more directories under your
gopher root which are named after your domain names. The primary vhost
directory (set with the -h <HOSTNAME> option) must exist or virtual
hosting will be disabled. Then simply add content to the hostname
directories and you're up and running.

Almost.

To make gopher clients work properly with virtual hosting, create a
root gophermap for each of your domains and include the "%" type
character to create a list of all available virtual hosts (see
README.gophermap). The generated virtual host links will be created so
that standard gopher clients will find the correct domain even when
they don't specifically tell the server which host they're trying to
reach.


CGI support
===========

uGopherServer supports most parts of the CGI/1.1 standard. Most
standard CGI variables are set, and some non-standard ones are added.

Currently, user input is broken due to debugging work done on the CGI
support. Query strings still work.

By default all scripts and binaries under any directory called
/cgi-bin/ are executed as CGI scripts (this includes cgi-bin
directories under users personal gopherspaces). Also, if a gophermap
is marked executable it is also processed as an CGI script.

As with regular files, CGI scripts must be world-executable (and
readable) or they will be ignored. Make sure your CGI script is safe
with ANY user input as poorly coded CGI scripts are the number #1
security problem with publicly open *nix servers.


Output filtering and PHP support
================================

In addition to CGI scripts uGopherServer supports output filtering
scripts. By default output filtering is turned off, but you can turn
it on by using the -f <FILTERDIR> option, creating that directory
and creating one or more scripts in there named by either the file
suffix, or by the gopher filetype char.

If a file is to be served out which matches either the file suffix
script, or the filetype script then instead of simply sending the
file to client the output filter script is executed with the
original file as the first parameter and the output of the script
is then sent to client.

For PHP support install the CLI version of the PHP interpreter and
then symlink (or copy) that binary to the directory specified with
-f option using the destination name "php".

$ ln -s /usr/bin/php5-cli /usr/lib/gophernicus/filters/php

After that all files with the php suffix will be "filtered" through
the PHP command line interpreter. In other words, PHP starts working.
And don't use the CGI version of PHP as it outputs HTTP headers the
gopher protocol doesn't have.


Charset support and conversions
===============================

uGopherServer supports three charsets: US-ASCII, ISO-8859-1 and
UTF-8. All textual input is internally upconverted to UTF-8 and then
downconverted to whatever charset the client is asking for. The
conversion is input autosensing which means that you don't have to
specify your filesystem charset, or the charset of your text files -
it's all detected automatically.

With standard gopher clients this is a bit of a problem as your text
files WILL be converted to 7-bit US-ASCII. This means that all 8-bit
charaters WILL BE LOST. This decision was made because no gopher
client that I tested was reliably cabable of decoding anything else
than pure US-ASCII. If you want to disable the conversion use the
"-no" option, or if you'd like to change the default output charset to
something else than US-ASCII just use for example the "-o ISO-8859-1"
option.


Selector rewriting
==================

Selector rewriting lets you rewrite parts of the selector on the fly.
Well, not parts, but really just the start of it. And the rewrite
engine here is nothing like Apache's mod_rewrite as I was too lazy
to integrate any regex libraries... So, all it does is rewrite a
fixed string at the start of the selector to something else. This
will let you move your directories around while making sure that
existing deeplinks still work.

Examples:

  -R "/~user=/~luser"
  -R "/old-dir=/new-dir"

*yawn*... Either this feature needs to be removed in Universal or
be drastically modified... Maybe send the rewriting to a CGI...


Session tracking and statistics
===============================

To enable virtual hosting with gopher (RFC1436) clients Gophernicus
tracks users and their session. As a side effect of that session
tracking, Gophernicus has simple throttling controls to keep nasty
users from killing your precious 120MHz PPC 604e server from dying
under the load. The throttling defaults are high enough that normal
human users will never hit the limits, but it's possible (and mostly
preferrable) that a badly behaving crawling agent will be throttled.

The current sessions and other real-time status data can be viewed
by opening the URL gopher://<HOSTNAME>/0/server-status . This status
view has been modeled after the Apache server-status which means
that it's possible to integrate Gophernicus into existing server
monitoring systems. To ease up such integrations, Gophernicus
supports HTTP requests of the server-status page using an URL like
http://<HOSTNAME>:70/server-status?auto




ChangeLog
[ Automatically generated from git log ]

2019-05-25  Ellenor Malik  <ellenor@umbrellix.net>

	* Added opportunistic SSL - if a misinformed Gopher hits the SSL port plaintext, it will get a plaintext menu without the 100k convention.

2018-02-28  Ellenor Malik  <ellenor@umbrellix.net>

	* Add auto-port-raise. You shouldn't need this if you configure your S/Gopher properly, but you might.

2018-02-27  Ellenor Malik  <ellenor@umbrellix.net>

	* yep, this is the big 1.0
	* S/Gopher proxy now in generic working order, and so is the server's SSL support. If you get any syslog cruft or worse, let me know.
	* I think it'll work now...
	* more to debug... fun
	* grr... let's see...
	* Or not.
	* Slower SSL support may actually be better, study says

2018-02-26  Ellenor Malik  <ellenor@umbrellix.net>

	* Rebranding almost complete.
	* Postamble gophermaps now work for directories. *doof*
	* Massive rebranding
	* First evidence of the new name of this Gopherd (uGopherServer, the universal gopher server) in a bundled app. Proxy is considered fully released.
	* The long awaited, the highly exalted, the first release of the S/Gopher proxy. Usage information is available at gopher://gopher.umbrellix.net/0/darknet-access/sgopher.doc
	* still doesn't notify of closure for CGIs, but eh
	* Should only do one thing now...
	* A bit of overkill here.... oh well, it's sorted out now I think.
	* well at least CGIs return now

2018-02-25  Ellenor Malik  <ellenor@umbrellix.net>

	* ngh... shutdown... the bad... connection...
	* ngh... shutdown... the bad... connection...
	* one... more... byte... and one... more... exit...
	* one... more... byte... and one... more... exit...
	* one... more... byte...
	* Grr, missing a character every fucking time
	* Fixed CGI support to behave properly in the presence of TLS.
	* minor change
	* SSL support works, as does plaintext support

2018-02-23  Ellenor Malik  <ellenor@umbrellix.net>

	* stunnel support complete. it was my fault it installed funny.
	* debbuging in production is pretty fun
	* debbuging in production is pretty fun
	* quick fix
	* stuff isn't great.
	* ah, compilation errors up on clang
	* SSL support doesn't work. However, you can use Stunnel if you want.
	* (Untested) SSL support now available. Note that its use is REQUIRED.

2018-02-21  Ellenor Malik  <ellenor@umbrellix.net>

	* Massively modified the overall structure; seems to be a pretty great piece of kit for a gopher server that wants inline maps.

2014-12-15  James Mills  <prologic@shortcircuit.net.au>

	* Improve counter example

2014-11-10  Kim Holviala  <kim@holviala.com>

	* Added MacOSX 10.10 to the list of supported systems in INSTALL
	* Released version 1.6
	* Fix compiling for MacOS X Yosemite

2014-05-15  Kim Holviala  <kim@holviala.com>

	* Updated TODO for systemd/upstart
	* Fix platform detection for Slackware (and possibly others)

2014-02-01  Kim Holviala  <kim@holviala.com>

	* Add license header to bin2c.c
	* Up version number to 1.6-whatever

2014-01-25  Kim Holviala  <kim@holviala.com>

	* Remove 'only tested with GCC' from INSTALL
	* Automatically copy ChangeLog to distribution dir on release
	* Automatically copy INSTALL & README to distribution dir on release
	* Add deb packaging leftovers to .gitignore
	* Up version number to 1.5 + include forgotten file git2changelog
	* Don't update ChangeLog unless the source tree is a git repo
	* Make ChangeLog update itself automatically from git log
	* Up Debian package version to 1.5
	* Added .gitignore
	* Clean up supported platforms list
	* Up version number to 1.5-rc1 (I mean really, this time)
	* Up version number to 1.5-rc1 & update copyright years to 2014
	* Search strings were being urldecoded against the specs (bugreport from Kacper Gutowski)
	* Fixed a (non-security) bug in strcut()
	* Initial git commit with version 1.4




ugopherserver.h
/*
 * Gophernicus - Copyright (c) 2009-2014 Kim Holviala <kim@holviala.com>
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef _GOPHERNICUS_H
#define _GOPHERNICUS_H


/*
 * Features
 */
#undef  ENABLE_STRICT_RFC1436	/* Follow RFC1436 to the letter */
#undef  ENABLE_AUTOHIDING	/* Hide manually listed resources from generated menus */


/*
 * Platform configuration
 */

/* Defaults should fit standard POSIX systems */
#define HAVE_IPv4		/* IPv4 should work anywhere */
#define HAVE_IPv6		/* Requires modern POSIX */
#define HAVE_PASSWD		/* For systems with passwd-like userdb */
#define PASSWD_MIN_UID 100	/* Minimum allowed UID for ~userdirs */
#define HAVE_LOCALES		/* setlocale() and friends */
#define HAVE_SHMEM		/* Shared memory support */
#define HAVE_UNAME		/* uname() */
#define HAVE_POPEN		/* popen() */
#undef  HAVE_STRLCPY		/* strlcpy() from OpenBSD */
#undef  HAVE_SENDFILE		/* sendfile() in Linux & others */

/* Linux */
#ifdef __linux
#undef  PASSWD_MIN_UID
#define PASSWD_MIN_UID 500
#define HAVE_SENDFILE
#endif

/* Embedded Linux with uClibc */
#ifdef __UCLIBC__
#undef HAVE_SHMEM
#undef HAVE_PASSWD
#endif

/* Haiku */
#ifdef __HAIKU__
#undef HAVE_SHMEM
#undef HAVE_PASSWD
#endif

/* OpenBSD */
#ifdef __OpenBSD__
#define HAVE_STRLCPY
#endif

/* MacOS */
#if defined(__APPLE__)
#if __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_10_10
#define HAVE_STRLCPY
#endif
#endif

/* Add other OS-specific defines here */

/*
 * Include headers
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <string.h>
#include <libgen.h>
#include <time.h>
#include <syslog.h>
#include <errno.h>
#include <pwd.h>
#include <limits.h>

#ifdef HAVE_SENDFILE
#include <sys/sendfile.h>
#include <fcntl.h>
#endif

#ifdef HAVE_LOCALES
#include <locale.h>
#endif

#ifdef HAVE_SHMEM
#include <sys/ipc.h>
#include <sys/shm.h>
#else
#define shm_state void
#endif

#if defined(HAVE_IPv4) || defined(HAVE_IPv6)
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#endif

#ifdef HAVE_UNAME
#include <sys/utsname.h>
#endif

#if !defined(HAVE_STRLCPY)
size_t strlcpy(char *dst, const char *src, size_t siz);
size_t strlcat(char *dst, const char *src, size_t siz);
#endif

/*
 * Compile-time configuration
 */

/* Common stuff */
#define CRLF		"\r\n"
#define EMPTY		""
#define PARENT		".."
#define ROOT		"/"

#define FALSE		0
#define TRUE		1

#define QUIT		1
#define OK		0
#define ERROR		-1

#define MATCH		0

/* Gopher filetypes */
#define TYPE_TEXT	'0'
#define TYPE_MENU	'1'
#define TYPE_ERROR	'3'
#define TYPE_GZIP	'5'
#define TYPE_QUERY	'7'
#define TYPE_BINARY	'9'
#define TYPE_GIF	'g'
#define TYPE_HTML	'h'
#define TYPE_INFO	'i'
#define TYPE_IMAGE	'I'
#define TYPE_MIME	'M'
#define TYPE_DOC	'd'
#define TYPE_TITLE	'!'

/* Protocols */
#define PROTO_GOPHER	'g'
#define PROTO_HTTP	'h'

/* Charsets */
#define AUTO		0
#define US_ASCII	1
#define ISO_8859_1	2
#define UTF_8		3

/* HTTP protocol stuff for logging */
#define HTTP_OK		200
#define HTTP_404	404
#define HTTP_DATE	"%d/%b/%Y:%T %z"
#define HTTP_USERAGENT	"Unknown gopher client"

/* Defaults for settings */
#define DEFAULT_HOST	"localhost"
#define DEFAULT_PORT	70
#define DEFAULT_TYPE	TYPE_TEXT
#define DEFAULT_MAP	"gophermap"
#define DEFAULT_TAG	"gophertag"
#define DEFAULT_HDR_EXT	"Ghdr"
#define DEFAULT_TAG_EXT	"Gtag"
#define DEFAULT_FTR_EXT	"Gftr"
#define DEFAULT_GPH_EXT	"gph"
#define DEFAULT_GMP_EXT	"gmp"
#define DEFAULT_SSL_CKF	"/m"
#define DEFAULT_CIPHERS	"HIGH:DEFAULT:!aNULL:!PSK:!SRP:!MD5:!RC4:!EXP"
#define DEFAULT_CGI	"/cgi-bin/"
#define DEFAULT_USERDIR	"public_gopher"
#define DEFAULT_ADDR	"unknown"
#define DEFAULT_WIDTH	70
#define DEFAULT_CHARSET	US_ASCII
#define MIN_WIDTH	33
#define MAX_WIDTH	600

/* Session defaults */
#define DEFAULT_SESSION_TIMEOUT		1800
#define DEFAULT_SESSION_MAX_KBYTES	4194304
#define DEFAULT_SESSION_MAX_HITS	4096

/* Dummy values for gopher protocol */
#define DUMMY_SELECTOR	"null"
#define DUMMY_HOST	"null.host\t1"

/* Safe $PATH for exec() */
#ifdef __HAIKU__
#define SAFE_PATH	"/boot/common/bin:/bin"
#else
#define SAFE_PATH	"/opt/local/bin:/usr/local/bin:/usr/bin:/bin"
#endif

/* Special requests */
#define SERVER_STATUS	"/server-status"
#define CAPS_TXT	"/caps.txt"

/* Error messages */
#define ERR_ACCESS	"Access denied!"
#define ERR_NOTFOUND	"File or directory not found!"

#define ERROR_HOST	"error.host\t1"
#define ERROR_PREFIX	"Error: "

/* Strings */
#define SERVER_SOFTWARE	"uGopherServer"
#define SERVER_SOFTWARE_FULL SERVER_SOFTWARE "-" VERSION

#define HEADER_FORMAT	"[%s]"
#define FOOTER_FORMAT	"Ecstatically rendered by UniversalGopherServer " VERSION " on %s"

#define UNITS		"KB", "MB", "GB", "TB", "PB", NULL
#define DATE_FORMAT	"%Y-%b-%d %H:%M"	/* See man 3 strftime */
#define DATE_WIDTH	17
#define DATE_LOCALE 	"POSIX"

#define USERDIR_FORMAT	"~%s", pwd->pw_name	/* See man 3 getpwent */
#define VHOST_FORMAT	"gopher://%s/"

/* ISO-8859-1 to US-ASCII look-alike conversion table */
#define ASCII \
	"E?,f..++^%S<??Z?" \
	"?''\"\"*--~?s>??zY" \
	" !c_*Y|$\"C?<?-R-" \
	"??23'u?*,1?>????" \
	"AAAAAAACEEEEIIII" \
	"DNOOOOO*OUUUUYTB" \
	"aaaaaaaceeeeiiii" \
	"dnooooo/ouuuuyty"

#define UNKNOWN '?'

/* Sizes & maximums */
#define BUFSIZE		(4*1024)	/* Default size for string buffers */
#define MAX_HIDDEN	32	/* Maximum number of hidden files */
#define MAX_FILETYPES	128	/* Maximum number of suffix to filetype mappings */
#define MAX_FILTERS	64	/* Maximum number of file filters */
#define MAX_SDIRENT	1024	/* Maximum number of files per directory to handle */
#define MAX_REWRITE	32	/* Maximum number of selector rewrite options */


// using one buffer now
extern char sockbuf[BUFSIZE*4];

/* Struct for file suffix -> gopher filetype mapping */
typedef struct {
	char suffix[15];
	char type;
} ftype;

/* Struct for selector rewriting */
typedef struct {
	char match[BUFSIZE];
	char replace[BUFSIZE];
} srewrite;

/* Struct for keeping the current options & state */

typedef struct {
	void *sslh;
	void *sslctx;
	int rfd;
	int wfd;
} sockstate;

typedef struct {

	/* Request */
	char req_selector[BUFSIZE];
	char req_realpath[BUFSIZE*6];
	char req_query_string[BUFSIZE];
	char req_referrer[BUFSIZE];
	char protection_certkeyfile[BUFSIZE];
	char protection_cipherlist[BUFSIZE];
	char req_local_addr[64];
	char req_remote_addr[64];
	char req_filetype;
	char req_protocol;
	off_t req_filesize;
	uid_t drop_uid;
	gid_t drop_gid;

	/* Output */
	int out_width;
	int out_charset;
	int out_protection;

	/* Settings */
	char server_description[64];
	char server_location[64];
	char server_platform[64];
	char server_admin[64];
	char server_root[256];
	char server_host_default[64];
	char server_host[64];
	int  server_port;

	char default_filetype;
	char map_file[64];
	char tag_file[64];
	char hdr_ext[64];
	char tag_ext[64];
	char ftr_ext[64];
	char cgi_file[64];
	char user_dir[64];
	char log_file[256];

	char hidden[MAX_HIDDEN][256];
	int hidden_count;

	ftype filetype[MAX_FILETYPES];
	int filetype_count;
	char filter_dir[64];

	srewrite rewrite[MAX_REWRITE];
	int rewrite_count;

	/* Session */
	int session_timeout;
	int session_max_kbytes;
	int session_max_hits;

	int (*read)(void *, char *, size_t);
	int (*write)(void *, const char *, size_t);
	char *(*fgets)(char *, size_t, void *);

	/* Feature options */
	char opt_parent;
	char opt_header;
	char opt_footer;
	char opt_date;
	char opt_syslog;
	char opt_magic;
	char opt_iconv;
	char opt_vhost;
	char opt_query;
	char opt_caps;
	char opt_shm;
	char opt_root;
	char debug;

	/* Preparing for future evented version of Gophernicus+Aster,
	 * socket state. Also used for SSL support, where available */
	sockstate ss;
} state;

/* Shared memory for session & accounting data */
#ifdef HAVE_SHMEM

#define SHM_KEY		0xbeeb0006	/* Unique identifier + struct version */
#define SHM_MODE	0600		/* Access mode for the shared memory */
#define SHM_SESSIONS	256		/* Max amount of user sessions to track */

typedef struct {
	long hits;
	long kbytes;

	time_t req_atime;
	char req_selector[128];
	char req_remote_addr[64];
	char req_filetype;

	char server_host[64];
	int  server_port;
} shm_session;

typedef struct {
	time_t start_time;
	long hits;
	long kbytes;
	char server_platform[64];
	char server_description[64];
	shm_session session[SHM_SESSIONS];
} shm_state;

#endif

/* Struct for directory sorting */
typedef struct {
	char	name[128];	/* Should be 256 but we're saving stack space */
	mode_t	mode;
	uid_t	uid;
	gid_t	gid;
	off_t	size;
	time_t	mtime;
} sdirent;


/* File suffix to gopher filetype mappings */
#define FILETYPES \
	"txt","0","pl","0","py","0","sh","0","tcl","0","c","0","cpp","0", "h","0","log","0", \
	"conf","0","php","0","php3","0", \
	"map","1","menu","1", \
	"hqx","4", \
	"Z","5","gz","5","tgz","5","tar","5","zip","5","bz2","5","rar","5","sea","5", \
	"q","7","qry","7", \
	"iso","9","so","9","o","9","rtf","9","ttf","9","bin","9", \
	"ics","c","ical","c", \
	"gif","g", \
	"html","h","htm","h","xhtml","h","css","h","swf","h","rdf","h","rss","h","xml","h", \
	"jpg","I","jpeg","I","png","I","bmp","I","svg","I","tif","I","tiff","I", \
	"ico","I","xbm","I","xpm","I","pcx","I", \
	"mbox","M", \
	"pdf","d","ps","d","doc","d","ppt","d","xls","d","xlsx","d","docx","d","pptx","d", \
	"mp3","s","wav","s","mid","s","wma","s","flac","s","ogg","s","aiff","s","aac","s", \
	"avi",";","mp4",";","mpg",";","mov",";","qt",";","asf",";","mpv",";","m4v",";", \
	NULL, NULL

/*
 * Useful macros
 */
#define strclear(str) str[0] = '\0';
#define sstrlcpy(dest, src) strlcpy(dest, src, sizeof(dest))
#define sstrlcat(dest, src) strlcat(dest, src, sizeof(dest))
#define sstrncmp(s1, s2) strncmp(s1, s2, sizeof(s2) - 1)
#define sstrncasecmp(s1, s2) strncasecmp(s1, s2, sizeof(s2) - 1)
#define sstrniconv(charset, out, in) strniconv(charset, out, in, sizeof(out))
#define max(a,b) (((a) > (b)) ? (a) : (b))
#define min(a,b) (((a) < (b)) ? (a) : (b))

/*
 * Include generated headers
 */
#include "functions.h"
#include "files.h"

#endif


